### ISSUE TITLE
Core.Page.ClientSideComponent properties  Id and Name have same guid value

### ISSUE THREAD

**Issue by kath561** [SharePoint/PnP-Sites-Core#1344]
#### Category
[ ]x Bug

#### Environment
[x ] Office 365 / SharePoint Online

#### Expected or Desired Behavior
change the Name property not to be a guid and to have the friendly text name

#### Observed Behavior
Both Name and Id properties contain the same  guid value

##### To Reproduce
cmdlet Get-PnPAvailableClientSideComponents at command pipeline position 1
Supply values for the following parameters:
Page: Home

Id                                   Name                                 
--                                   ----                                 
daf0b71c-6de8-4ef7-b511-faae7c388708 daf0b71c-6de8-4ef7-b511-faae7c388708 
e377ea37-9047-43b9-8cdb-a761be2f8e09 e377ea37-9047-43b9-8cdb-a761be2f8e09 
490d7c76-1824-45b2-9de3-676421c997fa 490d7c76-1824-45b2-9de3-676421c997fa 
b7dd04e1-19ce-4b24-9132-b60a1c2b910d b7dd04e1-19ce-4b24-9132-b60a1c2b910d 
20745d7d-8581-4a6c-bf26-68279bc123fc 20745d7d-8581-4a6c-bf26-68279bc123fc 

$cc3 = Get-PnPAvailableClientSideComponents -Page Home

examine $cc3 which is array of ClientSideComponent

Sharing is caring.

**Comment by ypcode** [comment]
Hi @kath561 

It is, indeed, the case, but it actually has noting to do with the PnP Core library.
Simply the value of the field Name is the Id for the SharePoint built-in WebParts

With your custom solution, however, it has the name of your component
![pnp-ps](https://user-images.githubusercontent.com/8345556/31017864-45ac067c-a52a-11e7-94f3-0bf50302629f.png)

Notice that it has now also "Alias" (but maybe only for First Release, I'm not sure)

For provisionning SharePoint OTB WebParts instead, you can use the -DefaultWebPartType argument instead that allows you through TAB or CTRL+SPACE to have intellisense
![pnp-ps2](https://user-images.githubusercontent.com/8345556/31017959-a44f9d10-a52a-11e7-969f-0ec46e0009a4.png)

If you need to get all available values from PS you can use:
`[OfficeDevPnP.Core.Pages.DefaultClientSideWebParts].GetEnumValues()`

Hope that helps.
Yannick

**Comment by kath561** [comment]
Unfortuneately it doesn't really help but rather shrugs off the problem. I am not a super smart programmer like many but I don't think this issue deserves to be kicked aside as just my incompetence.
I do think this is a real flaw in the command and makes it very difficult to use in automation projects. 

Putting an 'id' in a 'name' property is pretty confusing when you already have an Id field. I have to say I don't understand your answer as I don't have a custom solution but am referring to the teams sample solution that shows this command producing the friendly name.
The Alias you refer to only appears in console output and is not available as an object property so is pretty useless for automation.

There seems to be an assumption here that the command would only be used manually with intellisense etc rather than the focus of powershell which is automation.

 I would say that if the design intent is to have only guids then the property names should reflect this and not use the 'Name' when that is not what the property contains. It seems pointless to have 2 properties containing the same value. Also the documentation should show how that guid can be programmatically decoded to obtain the friendly name.

**Comment by ypcode** [comment]
Dear @kath561,

Don't get me wrong, I definitely did not want to kick anything aside and I'm sorry if my answer made you think otherwise.

I pretty much agree with you about this inconsistency, and I'm not sure why there is a Guid in place of name for the Default WebParts, but I am sure there is a good reason.
I guess it is related to localization concerns, a GUID translated to the term in the proper language. 
Maybe @jansenbe or @VesaJuvonen can tell us more about it.
Bottomline is, that's how it is persisted in SharePoint, the PnP Core lib only returns the data as it is

The Cmdlet returns the list of ClientSideComponent objects, so it can eventually be used with the .NET CSOM API through PowerShell.
I would think it would complexify the whole usage to wrap it into a custom object. Kinda same idea about "resolving the name" as you say, I think the objects should be returned as is and not altered.

About your suggestion of a parameter for Built-in or Custom, It might be a good idea, I will see what I could do towards this direction.
I will also investigate a bit more the Alias "property" I don't know very well.

About the documentation, It is automatically generated and displays the .NET type of the property. Indeed, it doesn't show a list of enum available values. 
@erwinvanhunen Is there a way to manually "fix" documentation? or maybe a change in the tool that would allow to display the enum values ?

About the intellisense thing, for sure, I was not suggesting you use this kind of thing for real automated processes, but more as a hint to know the available values for this parameter in the case you were searching a specific one.

In the mean time of getting something that will better suit your needs, I can give you this little snippet that will seamlessly gives you the real name of the WebPart or the Enum value when it is a Default WebPart

```
Function EnsureClientSideComponentName($componentName) {
    $defaultWpType = [OfficeDevPnP.Core.Pages.ClientSidePage]::NameToClientSideWebPartEnum($componentName)
    If ($defaultWpType -eq [OfficeDevPnP.Core.Pages.DefaultClientSideWebParts]::ThirdParty) {
        Return $componentName;
    }
    Else {
        Return $defaultWpType.ToString()
    }
}

# Usage
# With a built-in WP
Get-PnPAvailableClientSideComponents -Page Home | ?{"BingMap" -eq (EnsureClientSideComponentName($_.Name))  }
# With a custom WP
Get-PnPAvailableClientSideComponents -Page Home | ?{"KanbanBoard" -eq (EnsureClientSideComponentName($_.Name))  }
```

Hope that will help,

Best regards,
Yannick

**Comment by jansenbe** [comment]
Some quick feedback: the guid is because that's how 1st party web parts are handled internally. PS documentation will be refreshed once a month with our releases, expect some update next week.

**Comment by jansenbe** [comment]
and thanks @ypcode for the detailed answers!!
